src/DescriptorList.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include "Descriptor.hpp"
+#include "merc.h"
+
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <unordered_map>
+
+class DescriptorList {
+    // We hand out pointers to Descriptors, so this structure must be address-stable.
+    std::unordered_map<uint32_t, Descriptor> descriptors_;
+
+public:
+    // Return all descriptors, including those in the process of logging in.
+    [[nodiscard]] auto all() const noexcept {
+        return descriptors_ | ranges::views::transform([](auto &pr) -> const Descriptor & { return pr.second; });
+    }
+    [[nodiscard]] auto all() noexcept {
+        return descriptors_ | ranges::views::transform([](auto &pr) -> Descriptor & { return pr.second; });
+    }
+
+    // Return all descriptors for characters in the "playing" state.
+    [[nodiscard]] auto playing() const noexcept {
+        return all() | ranges::views::filter([](const Descriptor &d) { return d.is_playing(); });
+    }
+
+    // Return all descriptors for playing characters, skipping the given character.
+    [[nodiscard]] auto all_but(const CHAR_DATA *ch) const noexcept {
```

Comment:
pedantic, but all_but could also be all_playing_but  to distinguish it from the vanilla alls.